typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -727,6 +757,270 @@
        [_
         (fail #:reason "contract generation not supported for this type")]))))
 
+;; TODO full tests
+(define (type->static-contract/transient type)
+  (define typed-side 'both)
+  (let loop ([type type])
+    (define t->sc loop)
+    (define (prop->sc p)
+      ;;bg copied from above, but uses different t->sc
+      (match p
+        [(TypeProp: o (app t->sc sc))
+         (cond
+           [(not (equal? flat-sym (get-max-contract-kind sc)))
+            (raise-user-error 'type->static-contract/transient "proposition contract generation not supported for non-flat types")]
+           [else (is-flat-type/sc (obj->sc o) sc)])]
+        [(NotTypeProp: o (app t->sc sc))
+         (cond
+           [(not (equal? flat-sym (get-max-contract-kind sc)))
+            (raise-user-error 'type->static-contract/transient "proposition contract generation not supported for non-flat types")]
+           [else (not-flat-type/sc (obj->sc o) sc)])]
+        [(LeqProp: (app obj->sc lhs) (app obj->sc rhs))
+         (leq/sc lhs rhs)]
+        [(AndProp: ps)
+         (and-prop/sc (map prop->sc ps))]
+        [(OrProp: ps)
+         (or-prop/sc (map prop->sc ps))]))
+    (match type
+     ;; Implicit recursive aliases
+     [(Name: name-id args #f)
+      any/sc]
+     ;; Ordinary type applications or struct type names, just resolve
+     [(or (App: _ _)
+          (Name/struct:))
+      (t->sc (resolve-once type))]
+     [(Univ:) any/sc]
+     [(Bottom:) (or/sc)]
+     ;; This comes before Base-ctc to use the Value-style logic
+     ;; for the singleton base types (e.g. -Null, 1, etc)
+     [(Val-able: v)
+      (cond
+       [(eof-object? v)
+        (flat/sc #'eof-object?)]
+       [(void? v)
+        (flat/sc #'void?)]
+       [(or (symbol? v) (boolean? v) (keyword? v) (null? v) (eq? unsafe-undefined v))
+        (flat/sc #`(lambda (x) (eq? x '#,v)))]
```

Comment:
@samth FYI for my dissertation, this function is `(lambda (x) (void) (eq? x '#,v))` because of <https://github.com/racket/racket/issues/3339>

(If there's time to measure again later, I'll re-run without the `(void)` if possible)